ocaml/xapi/certs.ml

```diff
@@ -0,0 +1,414 @@
+
+  val parse_file_using : 'a Angstrom.t -> string -> ('a, string) result
+end = struct
+  open Angstrom
+
+  type t = {private_key: string; host_cert: string; other_certs: string list}
+
+  let is_whitespace = function
+    | ' ' ->
+        true
+    | '\t' ->
+        true
+    | '\n' ->
+        true
+    | '\r' ->
+        true
+    | _ ->
+        false
+
+  let is_newline = function '\n' -> true | '\r' -> true | _ -> false
+
+  let is_data = function '-' -> false | _ -> true
+
+  let newline = take_while1 is_newline
+
+  let ws = take_while is_whitespace
+
+  let data = take_while1 is_data
+
+  let key_header = string "-----BEGIN PRIVATE KEY-----"
+
+  let key_footer = string "-----END PRIVATE KEY-----"
+
+  let cert_header = string "-----BEGIN CERTIFICATE-----"
+
+  let cert_footer = string "-----END CERTIFICATE-----"
+
+  let crl_header = string "-----BEGIN X509 CRL-----"
+
+  let crl_footer = string "-----END X509 CRL-----"
+
+  let key =
+    ws *> key_header >>= fun hd ->
+    data >>= fun body ->
+    key_footer >>= fun tl -> ws *> return (String.concat "" [hd; body; tl])
+
+  let cert =
+    ws *> cert_header >>= fun hd ->
+    data >>= fun body ->
+    cert_footer >>= fun tl -> ws *> return (String.concat "" [hd; body; tl])
+
+  let host_pem =
+    key >>= fun private_key ->
+    cert >>= fun host_cert ->
+    many cert >>= fun other_certs ->
+    return {private_key; host_cert; other_certs}
+
+  let defer f = Fun.protect ~finally:f
+
+  let read_file path =
+    let ic = open_in path in
+    defer (fun () -> close_in ic) @@ fun () ->
+    really_input_string ic (in_channel_length ic)
+
+  let parse_string_using parser str =
+    let consume = Consume.Prefix in
+    parse_string ~consume parser str
+
+  let parse_file_using parser path =
+    try read_file path |> parse_string_using parser
+    with e ->
+      Error (Printf.sprintf "Can't process %s: %s" path (Printexc.to_string e))
+end
+
+(* module [Sync] provides an interface to certificates stored on
+disk. Certificates are read upon [init], potentially manipulated, and
+written back on [sync]. No other code should write these files. Writing
+back on [sync] provides a primitive form of transactions: in case of a
+fatal error, we could avoid writing a bad state to disk. This is
+currently not actively used, though *)
+
+module Sync : sig
+  type uuid = Uuidm.t
+
+  type paths = {
+      host_cert_path: string
+    ; ca_certs_dir: string
+    ; ca_crls_dir: string
+  }
+
+  type host = {
+      private_key: X509.Private_key.t
+    ; cert: X509.Certificate.t
+    ; intermediares: X509.Certificate.t list
+  }
+
+  type ca = {cert: X509.Certificate.t; uuid: uuid; mutable deleted: bool}
+
+  type cr = {name: string; cert: X509.Certificate.t; mutable deleted: bool}
+
+  (* [path] value to be used normally but other paths could be used for
+     testing *)
+  val default_paths : unit -> paths
+
+  (* initialise this module and read satte from disk *)
```

Comment:
I suspect `satte` is a German word :)

Reply:
Only as in "die satte Kuh" (the well-fed cow). I meant to use "state".